### PR TITLE
[GORDO-1599] Parallelize Pregel Graph Store

### PR DESCRIPTION
**Pull request by hkernbach**
### Scope & Purpose

Parallelize Pregel Graph Loading (GraphLoader)

#### Related Information

- [x] GitHub issue / Jira ticket: https://arangodb.atlassian.net/browse/GORDO-1599

### CODE REVIEW

File: arangod/Pregel/Worker/Worker.cpp [hunk 1]
```diff
@@ -594,8 +612,23 @@ auto Worker<V, E, M>::aqlResult(bool withId) const -> PregelResults {
...
+  if (!storeFutures.empty()) {
+    // wait for all futures to finalize
+    auto futureResponses = futures::collectAll(storeFutures).get();
+    for (auto const& fResponse : futureResponses) {
+      auto& result = fResponse.get();
+      if (result.fail()) {
+        LOG_PREGEL("ee2af", WARN) << fmt::format(
+            "Thread failed to store graph results: {}", result.errorMessage());
+        // TODO: Anything else we can do in case we've failed here?!
+        throw;
```
Comment (MBkkt) [inline arangod/Pregel/Worker/Worker.cpp:628]: throw what? `throw;` which is written below make sense only in catch block.
It's rethrow

File: arangod/Pregel/Worker/Worker.cpp [hunk 2]
```diff
@@ -594,8 +612,23 @@ auto Worker<V, E, M>::aqlResult(bool withId) const -> PregelResults {
   auto storer =
       GraphVPackBuilderStorer<V, E>(withId, _config, _algorithm->inputFormat(),
                                     std::move(_makeStatusCallback()));
-  // TODO GORDO-1599
-  storer.store(_quivers);
+  std::vector<futures::Future<Result>> storeFutures;
+  for (auto const& quiver : _quivers) {
+    storeFutures.emplace_back(storer.store({quiver}));
```
Comment (MBkkt) [inline arangod/Pregel/Worker/Worker.cpp:617]: Please do reserve

File: arangod/Pregel/Worker/Worker.cpp [hunk 3]
```diff
@@ -566,13 +566,31 @@ void Worker<V, E, M>::finalizeExecution(FinalizeExecution const& msg,
...
             auto storer = GraphStorer<V, E>(_config, _algorithm->inputFormat(),
                                             _config->globalShardIDs(),
                                             std::move(statusUpdateCallback));
             _feature.metrics()->pregelWorkersStoringNumber->fetch_add(1);
-            // TODO GORDO-1599
-            storer.store(_quivers);
+
+            std::vector<futures::Future<Result>> storeFutures;
+            for (auto const& quiver : _quivers) {
+              storeFutures.emplace_back(storer.store({quiver}));
```
Comment (MBkkt) [inline arangod/Pregel/Worker/Worker.cpp:578]: please do reserve

File: arangod/Pregel/Worker/Worker.cpp [hunk 4]
```diff
@@ -566,13 +566,31 @@ void Worker<V, E, M>::finalizeExecution(FinalizeExecution const& msg,
...
+              // wait for all futures to finalize
+              auto futureResponses = futures::collectAll(storeFutures).get();
+              for (auto const& fResponse : futureResponses) {
+                auto& result = fResponse.get();
+                if (result.fail()) {
+                  LOG_PREGEL("ee2af", WARN)
+                      << fmt::format("Thread failed to store graph results: {}",
+                                     result.errorMessage());
+                  // TODO: Anything else we can do in case we've failed here?!
+                  throw;
```
Comment (MBkkt) [inline arangod/Pregel/Worker/Worker.cpp:590]: throw what? `throw;` which is written below make sense only in catch block.
It's rethrow

File: arangod/Pregel/Worker/Worker.cpp [hunk 5]
```diff
@@ -594,8 +612,23 @@ auto Worker<V, E, M>::aqlResult(bool withId) const -> PregelResults {
...
+    storeFutures.emplace_back(storer.store({quiver}));
+  }
+  if (!storeFutures.empty()) {
+    // wait for all futures to finalize
+    auto futureResponses = futures::collectAll(storeFutures).get();
+    for (auto const& fResponse : futureResponses) {
+      auto& result = fResponse.get();
+      if (result.fail()) {
+        LOG_PREGEL("ee2af", WARN) << fmt::format(
+            "Thread failed to store graph results: {}", result.errorMessage());
```
Comment (MBkkt) [inline arangod/Pregel/Worker/Worker.cpp:626]: NIT Why?
```suggestion
        LOG_PREGEL("ee2af", WARN) <<
            "Thread failed to store graph results: " << result.errorMessage();
```

File: arangod/Pregel/GraphStore/GraphVPackBuilderStorer.h [hunk 6]
```diff
@@ -61,8 +61,9 @@ struct GraphVPackBuilderStorer : GraphStorerBase<V, E> {
         config(config),
         statusUpdateCallback(statusUpdateCallback) {}
 
+  // TODO: No sense to use a future here as it is implemented now...
```
Comment (MBkkt) [inline arangod/Pregel/GraphStore/GraphVPackBuilderStorer.h:64]: And why do you do it?

File: arangod/Pregel/Worker/Worker.cpp (hunk 3 above)
Comment (hkernbach) [reply arangod/Pregel/Worker/Worker.cpp:579]: added

File: arangod/Pregel/GraphStore/GraphVPackBuilderStorer.h (hunk 6 above)
Comment (hkernbach) [reply arangod/Pregel/GraphStore/GraphVPackBuilderStorer.h:64]: added

**Comment by markuspf** [comment]
This is made redundant by #18581

**Pull request closed without merging**